File: Week 6/sentimental-credit/credit.py
from cs50 import get_string


def main():
    cardNumber = get_string("Number: ")

    if (cardIsValid(cardNumber)):
        printCardBrand(cardNumber)
    else:
        print("INVALID")


def cardIsValid(cardNumber):
    numberSize = len(cardNumber)
    totalSum = 0

    for i in range(numberSize):
        currentDigit = int(cardNumber[numberSize - 1 - i])
        if i % 2 == 0:
            totalSum += currentDigit
        else:
            nTimesTwo = currentDigit * 2
            totalSum += ((nTimesTwo // 10) + (nTimesTwo % 10))

    if totalSum % 10 == 0:
        return True
    return False


def printCardBrand(cardNumber):

    cardLenght = len(cardNumber)
    firstTwoDigits = cardNumber[0:2]

    if (cardLenght == 15 and (firstTwoDigits in ["34", "37"])):
        print("AMEX")
    elif (cardLenght == 16 and (firstTwoDigits in ["51", "52", "53", "54", "55"])):
        print("MASTERCARD")
    elif (cardLenght in [13, 16] and (firstTwoDigits[0] == "4")):
        print("VISA")
    else:
        print("INVALID")


main()

File: Week 1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    int validateCardCheckSum(long cardNumber);
    int getLenghtOfCardNumber(long cardNumber);
    int getFirstTwoDigits(long cardNumber);
    long inputCardNumber, cardLenght, firstTwoDigits;

    inputCardNumber = get_long("Number: ");

    cardLenght = getLenghtOfCardNumber(inputCardNumber);
    firstTwoDigits = getFirstTwoDigits(inputCardNumber);

    if (validateCardCheckSum(inputCardNumber) == 0)
        printf("INVALID\n");
    else if (cardLenght == 15 && (firstTwoDigits == 34 || firstTwoDigits == 37))
        printf("AMEX\n");
    else if (cardLenght == 16 &&
             (firstTwoDigits == 51 || firstTwoDigits == 52 || firstTwoDigits == 53 || firstTwoDigits == 54 || firstTwoDigits == 55))
        printf("MASTERCARD\n");
    else if ((cardLenght == 13 || cardLenght == 16) && (firstTwoDigits / 10 == 4))
        printf("VISA\n");
    else
        printf("INVALID\n");
}

int validateCardCheckSum(long cardNumber)
{
    int totalAddition = 0, b = 0;

    for (int i = 0; i < 16; i++)
    {
        b = cardNumber % 10;

        if (i % 2 == 0)
        {
            totalAddition += b;
        }
        else
        {
            int nTimesTwo = b * 2;
            totalAddition += (((nTimesTwo) / 10) + ((nTimesTwo) % 10));
        }

        cardNumber /= 10;
    }

    if (totalAddition % 10 == 0)
    {
        return 1;
    }
    return 0;
}

int getLenghtOfCardNumber(long cardNumber)
{
    long size = 0;

    for (int i = 0; i < 16; i++)
    {
        if (cardNumber != 0)
        {
            size += 1;
        }
        cardNumber /= 10;
    }
    return size;
}

int getFirstTwoDigits(long cardNumber)
{
    long firstDigit = 0, secondDigit = 0;

    while (cardNumber != 0)
    {
        secondDigit = firstDigit;
        firstDigit = cardNumber % 10;
        cardNumber /= 10;
    }
    return (firstDigit * 10) + secondDigit;
}
